contrib/epee/src/wipeable_string.cpp

```diff
@@ -109,9 +115,112 @@ void wipeable_string::push_back(char c)
   buffer.back() = c;
 }
 
-void wipeable_string::pop_back()
+void wipeable_string::operator+=(char c)
+{
+  push_back(c);
+}
+
+void wipeable_string::operator+=(const std::string &s)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - s.size(), "Appended data too large");
+  grow(size(), size() + s.size());
+  const char *ptr = s.c_str();
+  while (*ptr)
+    push_back(*ptr++);
+}
+
+void wipeable_string::operator+=(const epee::wipeable_string &s)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - s.size(), "Appended data too large");
+  grow(size(), size() + s.size());
+  const char *ptr = s.data();
+  size_t sz = s.size();
+  while (sz--)
+    push_back(*ptr++);
+}
+
+void wipeable_string::operator+=(const char *s)
+{
+  size_t slen = strlen(s);
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - slen, "Appended data too large");
+  grow(size(), size() + slen);
+  while (*s)
+    push_back(*s++);
+}
+
+void wipeable_string::append(const char *data, size_t len)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - len, "Appended data too large");
+  grow(size(), size() + len);
+  while (len--)
```

Comment:
How is it "more obviously correct" ? Or safer? Are you implying that `std::memcpy` could copy the data to some other buffer? If so, document why you didn't use `std::memcpy` so that no one is tempted to change it. 

However, both implementations have the same issue - the compiler is free to do what it wants, provided the observable effects are the same.  `std::memcpy` has the advantage that its quicker to audit what the code wants to do, and the compiler automatically recognizes the call for optimizations. These `std::memcpy` optimizations are not proof that your implementation is more safe; C++ compilers will trigger auto-vectorization analysis in these types of loops. Its not uncommon to see compilers generate SSE code similar to `std::memcpy` in these situations, but with the downside that its duplicating ASM (and thus a small CPU cache usage increase). Or the compiler may even recognize that this is a `std::memcpy` and just what it normally does for `std::memcpy`.

If you turned down optimizations in this file in particular, then maybe it makes sense, but I would still be critical. Its difficult to control compiler output.

Reply:
Because the while version doesn't have to know at which offset to add the new data.